### PR TITLE
Add ETW tracing 

### LINKED ISSUE #2936: Add ETW tracing
### Description

See the existing CodeMarkers class

TODO
- [ ] decide what to record
- [ ] record it

? start/end: reloading settings from disc
? start/end: issue matching
? start/end: create execution context
? start/end: initialising the container
? fetching suppressions from the server

### PR DESCRIPTION
**Pull request by ugras-ergun-sonarsource**
fixes #2936

### CODE REVIEW

File: src/Integration/ETW/CodeMarkers.cs [hunk 1]
```diff
@@ -34,6 +34,7 @@ public static class Keywords
             public const EventKeywords Binding = (EventKeywords)2;
             public const EventKeywords Analysis = (EventKeywords)4;
             public const EventKeywords CFamily = (EventKeywords)8;
+            public const EventKeywords Suppression = (EventKeywords)2;
```
Comment (duncanp-sonar) [inline src/Integration/ETW/CodeMarkers.cs:37]: We're only recording Roslyn-related events for the moment, but we do handle suppressions for other languages. For now, I think making the key `RoslynSuppression` would be better as it will remind us in the future that it's Roslyn-specific.
The alternative is to add another keyword for `Roslyn`, but I don't think that is worth doing at the moment.

File: src/Roslyn.Suppressions/Roslyn.Suppressions/InProcess/RoslynSettingsFileSynchronizer.cs [hunk 2]
```diff
@@ -90,13 +91,18 @@ private void SuppressedIssuesMonitor_SuppressionsUpdateRequested(object sender,
             // does all of its work on a background thread.
             try
             {
+                CodeMarkers.Instance.FileSynchronizerUpdateRequesteStart();
```
Comment (duncanp-sonar) [inline src/Roslyn.Suppressions/Roslyn.Suppressions/InProcess/RoslynSettingsFileSynchronizer.cs:94]: This will tell us how often the update is requested. The actual update is happening on a different thread, and this method isn't waiting for the result so we don't know how long the update will take.

I'm not sure if it's worth recording Start/Stop if all of the work is being done on another thread, but I guess it doesn't hurt.

However, do we want to measure how long the actual update takes?

File: src/Roslyn.Suppressions/Roslyn.Suppressions/SonarDiagnosticSuppressor.cs [hunk 3]
```diff
@@ -70,7 +73,7 @@ public override void ReportSuppressions(SuppressionAnalysisContext context)
             {
                 return Enumerable.Empty<Suppression>();
             }
-
+            CodeMarkers.Instance.GetSuppressionsStart();
```
Comment (duncanp-sonar) [inline src/Roslyn.Suppressions/Roslyn.Suppressions/SonarDiagnosticSuppressor.cs:76]: Currently, this is going report as follows:
In standalone mode:
```
ReportSuppressionStart
ReportSuppressionStop
```

i.e. we have to infer that we are in standalone mode by the absence of the other events.

In connected mode:
```
ReportSuppressionStart
GetSuppressionsStart
GetSuppressionsStop(n)
ReportSuppressionStop
```
The two extra events are adding much from a timing point of view since they are called very close to the `ReportSuppressionStart/Stop`.

It would be simpler and more useful if we always reported something like:
```
ReportSuppressionStart
ReportSuppressionStop (mode, suppressionCount)
```

File: src/Integration/Integration.csproj [hunk 4]
```diff
@@ -76,6 +76,10 @@
     <ProjectReference Include="..\ProgressVS\ProgressVS.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Folder Include="ETW\" />
```
Comment (rita-sonarsource) [inline src/Integration/Integration.csproj:80]: Why is this needed?

File: src/Roslyn.Suppressions/Roslyn.Suppressions/SuppressionExecutionContext.cs [hunk 5]
```diff
@@ -55,5 +55,7 @@ private void GetProjectKey(AnalyzerOptions analyzerOptions)
         public bool IsInConnectedMode => SettingsKey != null;
 
         public string SettingsKey { get; private set; } = null;
+
+        public string Mode => SettingsKey != null ? "Connected" : "Standalone";
```
Comment (ugras-ergun-sonarsource) [inline src/Roslyn.Suppressions/Roslyn.Suppressions/SuppressionExecutionContext.cs:59]: Even it adds more complexity, I think having the mode name logic here makes the code more readable

File: src/Integration/Integration.csproj (hunk 4 above)
Comment (ugras-ergun-sonarsource) [reply src/Integration/Integration.csproj]: it's not. A leftover from moving file. Removed it.

File: src/Roslyn.Suppressions/Roslyn.Suppressions/SuppressionExecutionContext.cs (hunk 5 above)
Comment (rita-sonarsource) [inline src/Roslyn.Suppressions/Roslyn.Suppressions/SuppressionExecutionContext.cs:59]: Can we maybe do this conversion in the ETW method and pass along the boolean IsInConnectedMode? It feels odd to add this extra field to the class when it's used only for logging.

File: src/Roslyn.Suppressions/Roslyn.Suppressions/InProcess/RoslynSettingsFileSynchronizer.cs [hunk 6]
```diff
@@ -122,8 +122,7 @@ public async Task UpdateFileStorageAsync()
                 };
                 roslynSettingsFileStorage.Update(settings);
             }
-
-            logger.LogDebugExtended("End");
+            CodeMarkers.Instance.FileSynchronizerUpdateStop();
```
Comment (rita-sonarsource) [inline src/Roslyn.Suppressions/Roslyn.Suppressions/InProcess/RoslynSettingsFileSynchronizer.cs:125]: Just to point out that this won't be logged if we have an exception in the middle. I don't know if it's relevant from ETW's perspective to have the Stop when there are errors.

File: src/Core/ETW/CodeMarkers.cs [hunk 7]
```diff
@@ -34,6 +34,7 @@ public static class Keywords
             public const EventKeywords Binding = (EventKeywords)2;
             public const EventKeywords Analysis = (EventKeywords)4;
             public const EventKeywords CFamily = (EventKeywords)8;
+            public const EventKeywords RoslynSuppression = (EventKeywords)2;
```
Comment (rita-sonarsource) [inline src/Core/ETW/CodeMarkers.cs:37]: Isn't this supposed to be 16? why 2?

File: src/Core/ETW/CodeMarkers.cs [hunk 8]
```diff
@@ -118,6 +119,59 @@ public static class Keywords
...
+        private const int FileStorageUpdateStopId = 3007;
+
+        private const int FileWatcherInvalidateStartId = 3007;
+        private const int FileWatcherInvalidateStopId = 3008;
+
+        private const int FileSynchronizerUpdateStartId = 3009;
+        private const int FileSynchronizerUpdateStopId = 3010;
+
+
+        [Event(ReportSuppressionsStartId, Level = EventLevel.Informational, Keywords = Keywords.RoslynSuppression)]
```
Comment (rita-sonarsource) [inline src/Core/ETW/CodeMarkers.cs:143]: Is this still used?

File: src/Core/ETW/CodeMarkers.cs [hunk 9]
```diff
@@ -118,6 +119,59 @@ public static class Keywords
 
         #endregion
 
+        #region Roslyn Suppressions 3000-3999
+
+        private const int ReportSuppressionsStartId = 3000;
+        private const int ReportSuppressionsStopId = 3001;
+
+        private const int GetSuppressionsStartId = 3002;
```
Comment (rita-sonarsource) [inline src/Core/ETW/CodeMarkers.cs:127]: Some of these are not used anymore

File: src/Roslyn.Suppressions/Roslyn.Suppressions/SuppressionExecutionContext.cs (hunk 5 above)
Comment (ugras-ergun-sonarsource) [reply src/Roslyn.Suppressions/Roslyn.Suppressions/SuppressionExecutionContext.cs:59]: I thought of that, but I think this would make more sense from readability point and also we would not have context logic in ETW method.

Comment (rita-sonarsource) [reply src/Roslyn.Suppressions/Roslyn.Suppressions/SuppressionExecutionContext.cs:59]: Then just to make it readable maybe reuse `IsInConnectedMode` instead of checking the settings key again.

**Review by rita-sonarsource — CHANGES_REQUESTED** [review]
Please see my comments

File: src/Roslyn.Suppressions/Roslyn.Suppressions/InProcess/RoslynSettingsFileSynchronizer.cs (hunk 6 above)
Comment (duncanp-sonar) [reply src/Roslyn.Suppressions/Roslyn.Suppressions/InProcess/RoslynSettingsFileSynchronizer.cs:125]: I don't think ETW cares - it just logs whatever events it is given. I'm guessing that the matching of Start/Stop event pairs is done by PerfView.

We have try-finally blocks for the other event pairs, although for those the try-catch already existed. However, we may as well be consistent and have a try-finally so we always have matching pairs of events - it will make the logs easier to understand.

File: src/Roslyn.Suppressions/Roslyn.Suppressions/SonarDiagnosticSuppressor.cs [hunk 10]
```diff
@@ -52,6 +53,7 @@ internal SonarDiagnosticSuppressor(Func<IContainer> getContainer)
 
         public override void ReportSuppressions(SuppressionAnalysisContext context)
         {
+            CodeMarkers.Instance.ReportSuppressionsStart();
```
Comment (duncanp-sonar) [inline src/Roslyn.Suppressions/Roslyn.Suppressions/SonarDiagnosticSuppressor.cs:56]: This is now the only place we have ETW events that isn't in a try-finally block. However, I'd be inclined to leave it as it is.
I checked, and VS silently catches and squashes any exceptions thrown here (I thought it would report an AD0001 in the Error List, but it doesn't). 
If we see that we are losing "Stop" events we could add a try-finally at that point.

**Review by duncanp-sonar — APPROVED** [review]
